stateproof/builder.go

```diff
@@ -0,0 +1,470 @@
+	}
+	// validated that we can add the sig previously.
+	if err := builderForRound.Add(pos, sfa.Sig); err != nil {
+		// only Present called from Add returns an error which is already
+		// passed in the call above.
+		return network.Ignore, err
+	}
+	return network.Broadcast, nil
+}
+
+func (spw *Worker) builder(latest basics.Round) {
+	// We clock the building of state proofs based on new
+	// blocks.  This is because the acceptable state proof
+	// size grows over time, so that we aim to construct an extremely
+	// small state proof upfront, but if that doesn't work out, we
+	// will settle for a larger proof.  New blocks also tell us
+	// if a state proof has been committed, so that we can stop trying
+	// to build it.
+	for {
+		spw.tryBroadcast()
+
+		nextrnd := latest + 1
+		select {
+		case <-spw.ctx.Done():
+			spw.wg.Done()
+			return
+
+		case <-spw.ledger.Wait(nextrnd):
+			// Continue on
+		}
+
+		// See if any new state proofs were formed, according to
+		// the new block, which would mean we can clean up some builders.
+		hdr, err := spw.ledger.BlockHdr(nextrnd)
+		if err != nil {
+			spw.log.Warnf("spw.builder: BlockHdr(%d): %v", nextrnd, err)
+			continue
+		}
+
+		spw.deleteOldSigs(hdr)
+		spw.deleteOldBuilders(hdr)
+
+		// Broadcast signatures based on the previous block(s) that
+		// were agreed upon.  This ensures that, if we send a signature
+		// for block R, nodes will have already verified block R, because
+		// block R+1 has been formed.
+		proto := config.Consensus[hdr.CurrentProtocol]
+		newLatest := spw.ledger.Latest()
+		for r := latest; r < newLatest; r++ {
+			// Wait for the signer to catch up; mostly relevant in tests.
+			spw.waitForSignedBlock(r)
+
+			spw.broadcastSigs(r, proto)
+		}
+		latest = newLatest
+	}
+}
+
+// broadcastSigs periodically broadcasts pending signatures for rounds
+// that have not been able to form a state proof.
+//
+// Signature re-broadcasting happens in periods of proto.StateProofInterval
+// rounds.
+//
+// In the first half of each such period, signers of a block broadcast their
+// own signatures; this is the expected common path.
+//
+// In the second half of each such period, any signatures seen by this node
+// are broadcast.
+//
+// The broadcast schedule is randomized by the address of the block signer,
+// for load-balancing over time.
+func (spw *Worker) broadcastSigs(brnd basics.Round, proto config.ConsensusParams) {
+	if proto.StateProofInterval == 0 {
+		return
+	}
+
+	spw.mu.Lock()
+	defer spw.mu.Unlock()
+
+	var roundSigs map[basics.Round][]pendingSig
+	err := spw.db.Atomic(func(ctx context.Context, tx *sql.Tx) (err error) {
+		if brnd%basics.Round(proto.StateProofInterval) < basics.Round(proto.StateProofInterval/2) {
+			roundSigs, err = getPendingSigsFromThisNode(tx)
+		} else {
+			roundSigs, err = getPendingSigs(tx)
+		}
+		return
+	})
+	if err != nil {
+		spw.log.Warnf("broadcastSigs: getPendingSigs: %v", err)
+		return
+	}
+
+	for rnd, sigs := range roundSigs {
```

Comment:
looks like the lock is not really needed here in the loop since `spw.net` and `spw.log` are only used, and the Close function waits for this goroutine termination anyway.

Reply:
The lock is to prevent contention on the `spw.builders` (delete from the map while iterating over if).